### PR TITLE
Add disk usage in sidebar

### PR DESCRIPTION
**Pull request by yash-garg**
**Description**

- Clicking on the top left filebrowser logo takes back to the root directory
- Added disk usage of root path in sidebar with a progress bar below settings button (#1642)

image

:rotating_light: Before submitting your PR, please read [community](https://github.com/filebrowser/community), and indicate which issues (in any of the repos) are either fixed or closed by this PR. See [GitHub Help: Closing issues using keywords](https://help.github.com/articles/closing-issues-via-commit-messages/).

- [x] DO make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] DO make sure you are making a pull request against the **master branch** (left side). Also you should start *your branch* off *our master*.
- [x] DO make sure that File Browser can be successfully built. See [builds](https://github.com/filebrowser/community/blob/master/builds.md) and [development](https://github.com/filebrowser/community/blob/master/development.md).
- [x] DO make sure that related issues are opened in other repositories. I.e., the frontend, caddy plugins or the web page need to be updated accordingly.
- [x] AVOID breaking the continuous integration build.

**Further comments**
Tried to figure out and make it work as simply as possible, still not sure if this is the best possible solution. It'll be great if you guys can review and provide some feedback if it needs any kind of improvement!

### CODE REVIEW

**Comment by niubility000** [comment]
It looks pretty cool.

**Comment by niubility000** [comment]
I checked it out. I think it shows the space of the wrong partition. 
For example, in windows pc, filebrowser.exe is located in C:\, 
but my root is e:\   (assigned by the command: .\filebrowser -d filebrowser.db config set --root e:/) 
and the current user's root folder is e:\a (the value in the setting-user management--scope), 
It should show the space of partition where the current user's root folder is located. 
But currently it shows the space of c:

Another example is in linux system. The system disk is 10G,  and a mounted hard disk(/mnt/d ) is 4T. filebrowser is in the system disk, the root is /, but the current user's root folder is /mnt/d/aaa. 
It should show the space of /mnt/d.
But currently it shows the space of system disk.

**Comment by yash-garg** [comment]
> I checked it out. I think it shows the space of the wrong partition. For example, in windows pc, filebrowser.exe is located in C:, but my root is e:\ (assigned by the command: .\filebrowser -d filebrowser.db config set --root e:/) and the current user's root folder is e:\a (the value in the setting-user management--scope), It should show the space of partition where the current user's root folder is located. But currently it shows the space of c:
> 
> Another example is in linux system. The system disk is 10G, and a mounted hard disk(/mnt/d ) is 4T. filebrowser is in the system disk, the root is /, but the current user's root folder is /mnt/d/aaa. It should show the space of /mnt/d. But currently it shows the space of system disk.

My bad will have a look at it, probably I am using the wrong path. Thanks for the update!

**Comment by yash-garg** [comment]
@niubility000 Can u please check once again after pulling the changes? Should work fine now. Thanks!

**Comment by niubility000** [comment]
> @niubility000 Can u please check once again after pulling the changes? Should work fine now. Thanks!

I tested again. Now it shows the space of partition where the admin's root folder is located  (Assigned by .\filebrowser -d filebrowser.db config set --root e:/ ). 
It works perfect in windows. 
But in linux, there is a situation that the partition of current user's root folder may be different to the partition of the admin's root folder.  For example, the admin's  root folder is /, (the partition is system disk), and the current user's root folder is /mnt/d/aaa (assigned in "setting--user management--current user--scope", and then the partition is /mnt/d). I think it's better to show the space of the mounted disk /mnt/d. 

Another small bug is when you click setting, this space bar will disappear.

Maybe it's hard to get the current user's scope value from backend, so this version is good enough to use!

**Comment by yash-garg** [comment]
> > @niubility000 Can u please check once again after pulling the changes? Should work fine now. Thanks!
> 
> I tested again. Now it shows the space of partition where the admin's root folder is located (Assigned by .\filebrowser -d filebrowser.db config set --root e:/ ). It works perfect in windows. But in linux, there is a situation that the partition of current user's root folder may be different to the partition of the admin's root folder. For example, the admin's root folder is /, (the partition is system disk), and the current user's root folder is /mnt/d/aaa (assigned in "setting--user management--current user--scope", and then the partition is /mnt/d). I think it's better to show the space of the mounted disk /mnt/d.
> 
> Maybe it's hard to get the current user's scope value from backend, so this version is good enough to use!

Thanks for the review. Will have a look in Linux once and update if I manage to solve it!

**Comment by niubility000** [comment]
> Thanks for the review. Will have a look in Linux once and update if I manage to solve it!

Maybe a easier and better way is to show the space of partition where the current dir we are browsing is located. For example, I logged in as an admin, If I am browsing "/etc", then the space partition should be the system disk, If I am browsing /mnt/d/family, the space partition should be the /mnt/d. it's more accurate.

**Comment by yash-garg** [comment]
Sure, will have a look at it.

**Comment by niubility000** [comment]
Test this one in resource.go:
`usage := disk.GetDiskUsage(filepath.Dir(d.user.FullPath(r.URL.Path)))`
It works perfect!

and I think in sidebar.vue, just change the bar to this place is better:

`
.......
      </router-link>
    </template>

    <div class="credits"
      v-if="$router.currentRoute.path.includes('/files/')"
      style="width:90%; margin: 0em 2.5em 3em 2.5em;"
    >
      <progress-bar :val="req.diskUsedPercent" :size="large"></progress-bar>
      <br />
      {{ req.freeDiskSpace }} free of {{ req.totalDiskSpace }}
    </div>  

    <p class="credits">
      <span>
        <span v-if="disableExternal">File Browser</span>
        <a
          v-else
          rel="noopener noreferrer"
          target="_blank"
          href="https://github.com/filebrowser/filebrowser"
          >File Browser</a
        >
        <span> {{ version }}</span>
      </span>
      <span
        ><a @click="help">{{ $t("sidebar.help") }}</a></span
      >
    </p>
`

**Comment by yash-garg** [comment]
> Test this one in resource.go: `usage := disk.GetDiskUsage(filepath.Dir(d.user.FullPath(r.URL.Path)))` It works perfect now!
> 
> and I think in sidebar.vue, just change the bar to this place is better:
> 
> ` .......
> 
> ```
> <div class="credits"
>   v-if="$router.currentRoute.path.includes('/files/')"
>   style="width:90%; margin: 0em 2.5em 3em 2.5em;"
> >
>   <progress-bar :val="req.diskUsedPercent" :size="large"></progress-bar>
>   <br />
>   {{ req.freeDiskSpace }} free of {{ req.totalDiskSpace }}
> </div>  
> 
> <p class="credits">
>   <span>
>     <span v-if="disableExternal">File Browser</span>
>     <a
>       v-else
>       rel="noopener noreferrer"
>       target="_blank"
>       href="https://github.com/filebrowser/filebrowser"
>       >File Browser</a
>     >
>     <span> {{ version }}</span>
>   </span>
>   <span
>     ><a @click="help">{{ $t("sidebar.help") }}</a></span
>   >
> </p>
> ```
> 
> `

Will do, thanks!

**Comment by yash-garg** [comment]
Made the changes you mentioned, was busy these days so couldn't do earlier.

**Comment by hernandito** [comment]
Hi Guys,

I hate to hijack this thread... but seems like there is not much activity at the main branch. I want to do an edit per the FEATURE REQUEST I posted here:

[Feature Request - Folder and File Icon Differentiation](https://github.com/filebrowser/filebrowser/issues/1698)

I believe I have identified the source files where to make this change... But unfortunately I am not able to build my edits (with NPM and GO). I get errors galore in the many environments I have tried. I simply cannot get the "**filebrowser**" binary file created.

Does anyone have an idea/thoughts/comments? Please reply on the Feature Request issue linked above... I don't want to hijack this thread.

Thank you.

File: frontend/src/components/header/HeaderBar.vue [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <template>
   <header>
-    <img v-if="showLogo !== undefined" :src="logoURL" />
+    <a href="/">
+      <img v-if="showLogo !== undefined" :src="logoURL" alt="logo" />
+    </a>
```
Comment (o1egl) [inline frontend/src/components/header/HeaderBar.vue:5]: I think it should be
```suggestion
    <a v-if="showLogo !== undefined" href="/">
      <img :src="logoURL" alt="logo" />
    </a>
```

File: files/file.go [hunk 2]
```diff
@@ -26,20 +26,24 @@ import (
...
+	IsSymlink       bool              `json:"isSymlink"`
+	Type            string            `json:"type"`
+	Subtitles       []string          `json:"subtitles,omitempty"`
+	Content         string            `json:"content,omitempty"`
+	Checksums       map[string]string `json:"checksums,omitempty"`
+	Token           string            `json:"token,omitempty"`
+	DiskUsedPercent string            `json:"diskUsedPercent"`
+	FreeDiskSpace   string            `json:"freeDiskSpace"`
+	TotalDiskSpace  string            `json:"totalDiskSpace"`
+	UsedDiskSpace   string            `json:"usedDiskSpace"`
```
Comment (o1egl) [inline files/file.go:46]: Let's return values in bytes and render them on the client-side. Similar to the `Size` field

File: disk/usage.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+import "github.com/shirou/gopsutil/disk"
+
+func GetDiskUsage(path string) *disk.UsageStat {
+	usage, err := disk.Usage(path)
+	if err != nil {
+		return nil
+	}
+	return usage
+}
```
Comment (o1egl) [inline disk/usage.go:11]: Why do we need this wrapper?

File: files/file.go (hunk 2 above)
Comment (o1egl) [inline files/file.go:46]: I don't think that those values should be a part of the `FileInfo` struct. It should be extracted into a separate struct and fetched in a separate http query

**Review by o1egl — CHANGES_REQUESTED** [review]

**Comment by o1egl** [comment]
https://github.com/filebrowser/filebrowser/commit/d1d8e3e3405381b01317fe07ae729d70219415a7

**Pull request closed without merging**